HIP/hip-367.md

```diff
@@ -0,0 +1,238 @@
+- As a user I should be able to pay for as many token associations upfront as I want to, using the automatic association slots 
+- As a primary creator of an NFT collection, I should be able to mint a large number of NFT serials in a single collection
+- As a secondary NFT marketplace dApp, I should be able to create many thousands of collections (created by multiple primary creators of the collections), such that each collection can hold a large number of NFTs
+- As an avid collector of NFT, I want to manage my entire collection within the same account.
+
+## Specification
+Currently, we have a limit of 1000 token associations per Account put in place which is dictated by the bootstrap property `tokens.maxPerAccount`.
+This is enforced everytime we associate token/tokens to an account. 
+
+When using the HederaTokenStore :
+```java
+    public ResponseCodeEnum associate(AccountID aId, List<TokenID> tokens, boolean automaticAssociation) {
+      ...
+          if ((accountTokens.numAssociations() + tokenIds.size()) > properties.maxTokensPerAccount()) {
+              validity = TOKENS_PER_ACCOUNT_LIMIT_EXCEEDED;
+          }
+      ...
+    }
+```
+
+When using the Account model:
+```java
+    public void associateWith(List<Token> tokens, int maxAllowed, boolean automaticAssociation){
+        final var alreadyAssociated=associatedTokens.size();
+        final var proposedNewAssociations=tokens.size()+alreadyAssociated;
+        validateTrue(proposedNewAssociations<=maxAllowed,TOKENS_PER_ACCOUNT_LIMIT_EXCEEDED);
+        ...
+    }
+```
+
+This limit enforcement needs to be removed. This limit is also enforced when creating and updating an account with maxAutomaticAssociations described by this [HIP-23](https://github.com/anighanta/hedera-improvement-proposal/blob/master/HIP/hip-23.md)
+That enforcement also has to be removed.
+
+The list of tokens associated to an account are saved on the state using a MerkleLeaf node `MerkleAccountTokens` which holds the array of Ids of the token as `CopyOnWriteIds` and each MerkleAccount will have this as a child. 
+The new way of efficiently doing this is by using a `MapValueLinkedList` data structure which will be represented by Account-Token relationship Map of <EntityNumPair, MerkleTokenRelStatus> in which `MerkleTokenRelStatus` serves as a node of a doubly linkedList. 
+The `nextKey` and `prevkey` in this `MerkleTokenRelStatus`will point to the next and previous token relationships of the account. And `tokenAssociaitonMetadata` is saved on MerkleAccountState.
+The `TokenAssociationMetadata` encapsulates three properties of the MerkleAccount
+1. The total number of tokens that this account is associated with `numAssociations`
+2. The total number among these associations which have zero balance `numZeroBalances`
+3. And finally the latest association represented as an `EntityNumPair` of MerkleAccount's id and token's id `lastAssociation` 
```

Comment:
3. And finally the latest association represented as an `EntityNumPair` of account's id and token's id as `lastAssociation` 
```